### PR TITLE
core/consensus: add histogram metric for consensus duration

### PR DESCRIPTION
**Pull request by dB2510**
Adds a histogram metric to report consensus duration for each consensus instance.

category: feature
ticket: #1247

### CODE REVIEW

**Review by corverroos — APPROVED** [review]

**Review by corverroos — CHANGES_REQUESTED** [review]
This measures the incorrect thing

File: core/consensus/component.go [hunk 1]
```diff
@@ -54,7 +54,7 @@ func Protocols() []protocol.ID {
 type subscriber func(ctx context.Context, duty core.Duty, value proto.Message) error
 
 // newDefinition returns a qbft definition (this is constant across all consensus instances).
-func newDefinition(nodes int, subs func() []subscriber) qbft.Definition[core.Duty, [32]byte] {
+func newDefinition(nodes int, subs func() []subscriber, startTime time.Time) qbft.Definition[core.Duty, [32]byte] {
```
Comment (corverroos) [inline core/consensus/component.go:57]: read the `newDefinition` godoc, this still instruments the wrong thing.

**Review by corverroos — CHANGES_REQUESTED** [review]

File: core/consensus/component.go [hunk 2]
```diff
@@ -301,9 +299,20 @@ func (c *Component) propose(ctx context.Context, duty core.Duty, value proto.Mes
...
+		instrumentConsensus(duty, qcommit[0].Round(), t0)
+		c.def.Decide(ctx, duty, val, qcommit)
+	}
+
 	// Run the algo, blocking until the context is cancelled.
-	err = qbft.Run[core.Duty, [32]byte](ctx, c.def, qt, duty, peerIdx, hash)
-	if err != nil && !isContextErr(err) {
+	err = qbft.Run[core.Duty, [32]byte](ctx, def, qt, duty, peerIdx, hash)
+	if isContextErr(err) {
+		consensusTimeout.WithLabelValues(duty.String()).Inc()
```
Comment (corverroos) [inline core/consensus/component.go:313]: think about this one some more

Comment (dB2510) [reply core/consensus/component.go:313]: It times out when context is closed but we haven't DECIDED

**Review by corverroos — APPROVED** [review]

**Pull request merged**